tests/lib/linter/linter.js

```diff
@@ -349,3875 +284,3623 @@ describe("Linter", () => {
 
-                const node2 = context.getNodeByRangeIndex(3);
+            const alternateParser = "esprima";
 
-                assert.strictEqual(node2.type, "VariableDeclaration");
-                return {};
+            linter.defineParser("esprima", esprima);
+            linter.defineRule("test-rule", {
+                create: sinon.mock().withArgs(
+                    sinon.match({ parserPath: alternateParser })
+                ).returns({})
             });
 
-            linter.defineRule("checker", { create: spy });
-            linter.verify(code, config);
-            assert(spy.calledOnce);
+            const config = { rules: { "test-rule": 2 }, parser: alternateParser };
+
+            linter.verify("0", config, filename);
         });
 
-        it("should return null if the index is outside the range of any node", () => {
-            const config = { rules: { checker: "error" } };
-            const spy = sinon.spy(context => {
-                const node1 = context.getNodeByRangeIndex(-1);
+        it("should use parseForESLint() in custom parser when custom parser is specified", () => {
+            const config = { rules: {}, parser: "enhanced-parser" };
 
-                assert.isNull(node1);
+            linter.defineParser("enhanced-parser", testParsers.enhancedParser);
+            const messages = linter.verify("0", config, filename);
+            const suppressedMessages = linter.getSuppressedMessages();
 
-                const node2 = context.getNodeByRangeIndex(-99);
+            assert.strictEqual(messages.length, 0);
+            assert.strictEqual(suppressedMessages.length, 0);
+        });
 
-                assert.isNull(node2);
-                return {};
+        it("should expose parser services when using parseForESLint() and services are specified", () => {
+            linter.defineParser("enhanced-parser", testParsers.enhancedParser);
+            linter.defineRule("test-service-rule", {
+                create: context => ({
+                    Literal(node) {
+                        context.report({
+                            node,
+                            message: context.sourceCode.parserServices.test.getMessage()
+                        });
+                    }
+                })
             });
 
-            linter.defineRule("checker", { create: spy });
-            linter.verify(code, config);
-            assert(spy.calledOnce);
-        });
-    });
+            const config = { rules: { "test-service-rule": 2 }, parser: "enhanced-parser" };
+            const messages = linter.verify("0", config, filename);
+            const suppressedMessages = linter.getSuppressedMessages();
 
+            assert.strictEqual(messages.length, 1);
+            assert.strictEqual(messages[0].message, "Hi!");
+            assert.strictEqual(suppressedMessages.length, 0);
+        });
 
-    describe("when calling context.getScope", () => {
```

Comment:
I just noticed that this set of tests wasn't copied into source-code tests, but we could do that in another PR.